pkg/vpcmodel/nodesExplainability.go

```diff
@@ -25,22 +36,111 @@ func (c *VPCConfig) getVsiNode(name string) Node {
-func (c *VPCConfig) ExplainConnectivity(srcName, dstName string) (explanation string, err error) {
-	src := c.getVsiNode(srcName)
-	if src == nil {
-		return "", fmt.Errorf("src %v does not represent a VSI", srcName)
+// given input cidr, gets (disjoint) external nodes I s.t.:
+//  1. The union of these nodes is the cidr
+//  2. Let i be a node in I and n be a node in VPCConfig.
+//     i and n are either disjoint or i is contained in n
+//     Note that the vpconfig nodes were chosen w.r.t. connectivity rules (SG and NACL)
+//     s.t. each node either fully belongs to a rule or is disjoint to it.
+//     to get nodes I as above:
+//  1. Calculate the IP blocks of the nodes N
+//  2. Calculate from N and the cidr block, disjoint IP blocks
+//  3. Return the nodes created from each block from 2 contained in the input cidr
+func (c *VPCConfig) getCidrExternalNodes(cidr string) (cidrDisjointNodes []Node, err error) {
+	cidrsIPBlock := common.NewIPBlockFromCidr(cidr)
+	if cidrsIPBlock == nil { // string cidr does not represent a legal cidr
+		return nil, nil
+	}
+	cidrIPBlocks := []*common.IPBlock{cidrsIPBlock}
+	// 1.
+	vpcConfigNodesExternalBlock := make([]*common.IPBlock, 0)
+	for _, node := range c.Nodes {
+		if !node.IsExternal() {
+			continue
+		}
+		thisNodeBlock := common.NewIPBlockFromCidr(node.Cidr())
+		vpcConfigNodesExternalBlock = append(vpcConfigNodesExternalBlock, thisNodeBlock)
 	}
-	dst := c.getVsiNode(dstName)
-	if dst == nil {
-		return "", fmt.Errorf("dst %v does not represent a VSI", dstName)
+	// 2.
+	disjointBlocks := common.DisjointIPBlocks(cidrIPBlocks, vpcConfigNodesExternalBlock)
+	// 3.
+	cidrDisjointNodes = make([]Node, 0)
+	for _, block := range disjointBlocks {
+		if block.ContainedIn(cidrsIPBlock) {
+			node, err1 := newExternalNode(true, block)
+			if err1 != nil {
+				return nil, err1
+			}
+			cidrDisjointNodes = append(cidrDisjointNodes, node)
+		}
 	}
-	rulesOfConnection, err1 := c.GetRulesOfConnection(src, dst)
+	return cidrDisjointNodes, nil
+}
+
+// given a string or a vsi or a cidr returns the corresponding node(s)
+func (c *VPCConfig) getNodesFromInput(cidrOrName string) ([]Node, error) {
+	if vsi := c.getVsiNode(cidrOrName); vsi != nil {
+		return []Node{vsi}, nil
+	}
+	return c.getCidrExternalNodes(cidrOrName)
+}
+
+// todo: group results. for now just prints each
+
+// ExplainConnectivity todo: this will not be needed here once we connect explanbility to the cli
+func (c *VPCConfig) ExplainConnectivity(src, dst string) (explanation string, err error) {
+	explanationStruct, err1 := c.computeExplainRules(src, dst)
 	if err1 != nil {
 		return "", err1
 	}
-	return rulesOfConnection.String(src, dst, c)
+	err2 := explanationStruct.computeConnections(c)
+	if err2 != nil {
+		return "", err2
+	}
+	return explanationStruct.String(c)
+}
+
+func (c *VPCConfig) computeExplainRules(srcName, dstName string) (explanationStruct explainStruct, err error) {
+	srcNodes, dstNodes, err := c.processInput(srcName, dstName)
```

Comment:
please document function.
the trick with the 2D loop, while there is only 1D list, is confusing

Reply:
Documentation added
You are right regarding the 2D and 1D, but trying to avoid the 2D loop will be uglier. I added an explanation. See if it helps.